### PR TITLE
feat: [CI-11277]: Support Signed URLs for Cache Intelligence workflows

### PR DESCRIPTION
**Pull request by sahithibanda01**
Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our [code of conduct](CODE_OF_CONDUCT.md), and the process for submitting pull requests to us.

Fixes #

## Proposed Changes

-
-
-

### Checklist

- [ ] Read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [ ] Read the [**CODE OF CONDUCT**](CODE_OF_CONDUCT.md) document.
- [ ] Add tests to cover changes.
- [ ] Ensure your code follows the code style of this project.
- [ ] Ensure CI and all other PR checks are green OR
    - [ ] Code compiles correctly.
    - [ ] Created tests which fail without the change (if possible).
    - [ ] All new and existing tests passed.
- [ ] Add your changes to `Unreleased` section of [CHANGELOG](CHANGELOG.md).
- [ ] Improve and update the [README](README.md) (if necessary).
- [ ] Ensure [documentation](./DOCS.md) is up-to-date. The same file will be updated in [plugin index](https://github.com/drone/drone-plugin-index/blob/master/content/meltwater/drone-cache/index.md) when your PR is accepted, so it will be available for end-users at http://plugins.drone.io.

### CODE REVIEW

File: harness/http.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+func (c *HTTPClient) GetDownloadPresignURL(ctx context.Context, key string) (string, error) {
+	path := fmt.Sprintf(harnessRestoreLinkEndpoint, c.AccountID, key)
+	return c.getLink(ctx, c.Endpoint+path)
+}
+
+func (c *HTTPClient) GetExistsPresignURL(ctx context.Context, key string) (string, error) {
+	path := fmt.Sprintf(harnessExistsLinkEndpoint, c.AccountID, key)
+	return c.getLink(ctx, c.Endpoint+path)
+}
+
```
Comment (hemanthmantri) [inline harness/http.go:56]: Nit: Please add docstrings for all methods describing their use.

File: storage/backend/harness/harness.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+	return res.Header.Get("ETag") != "", nil
+}
+
+func (b *Backend) List(ctx context.Context, key string) ([]common.FileEntry, error) {
+	// not implemented
+	return nil, errors.New("list operation not implemented")
+}
+
+func (b *Backend) do(ctx context.Context, method, url string, body io.Reader) (*http.Response, error) {
```
Comment (hemanthmantri) [inline storage/backend/harness/harness.go:96]: Can we use a more descriptive name here (unless this is some convention)?

File: storage/backend/harness/config.go [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package harness
+
+// Config is a structure to store S3  backend configuration.
```
Comment (hemanthmantri) [inline storage/backend/harness/config.go:3]: Can we remove S3?

File: storage/backend/backend.go [hunk 4]
```diff
@@ -28,6 +29,8 @@ const (
 	S3 = "s3"
 	// SFTP type of the corresponding backend represented as string constant.
 	SFTP = "sftp"
+	//Harness type of the corresponding backend represented as string constant.
+	Harness = "harness"
```
Comment (hemanthmantri) [inline storage/backend/backend.go:33]: This drone plugin is open-source and in drone-plugins org. Is it okay to  implement something harness specific (which cannot be used by non-harness users)?  Do we have to fork the plugin in such cases? 
cc: @rutvijmehta-harness @vistaarjuneja

File: main.go [hunk 5]
```diff
@@ -504,6 +505,16 @@ func main() {
 			Usage:   "sftp port",
 			EnvVars: []string{"SFTP_PORT"},
 		},
+		&cli.StringFlag{
+			Name:    "cache-service-token",
+			Usage:   "cache service token",
+			EnvVars: []string{"PLUGIN_CACHE_SERVICE_TOKEN"},
```
Comment (rutvijmehta-harness) [inline main.go:511]:
```suggestion
			EnvVars: []string{"PLUGIN_CACHE_SERVICE_BEARER_TOKEN"},
```

File: storage/backend/backend.go (hunk 4 above)
Comment (rutvijmehta-harness) [reply storage/backend/backend.go:33]: This is a backward compatible change so it shouldn't break anything but we can get input from product.

File: storage/backend/harness/harness.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	defer internal.CloseWithErrLogf(b.logger, res.Body, "response body, close defer")
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("received status code %d from presigned put url", res.StatusCode)
+	}
+
+	return nil
+}
+
+func (b *Backend) Exists(ctx context.Context, key string) (bool, error) {
+	preSignedURL, err := b.client.GetExistsPresignURL(ctx, key)
```
Comment (rutvijmehta-harness) [inline storage/backend/harness/harness.go:73]: @hemanthmantri do we need presigned URL for `Exists` or the server can just return a boolean?

File: storage/backend/harness/harness.go (hunk 2 above)
Comment (sahithibanda01) [reply storage/backend/harness/harness.go:96]: hey hemanth this is the convention which we follow in all go repos

File: storage/backend/backend.go (hunk 4 above)
Comment (hemanthmantri) [reply storage/backend/backend.go:33]: Checked with @bradrydzewski and he confirmed it is okay to add Harness-specific code as long as is is backward compatible. Thanks.

**Review by rutvijmehta-harness — APPROVED** [review]

**Review by HenAmar — APPROVED** [review]

**Pull request merged**